src/main/java/com/o19s/es/termstat/TermStatQuery.java

```diff
@@ -0,0 +1,122 @@
+        this.expr = expr;
+        this.aggr = aggr;
+        this.posAggr = posAggr;
+        this.terms = terms;
+    }
+
+
+    public String getExpr() {
+        return this.expr;
+    }
+    public AggrType getAggr() { return this.aggr; }
+    public AggrType getPosAggr() { return this.posAggr; }
+    public Set<Term> getTerms() { return this.terms; }
+
+
+    @SuppressWarnings("EqualsWhichDoesntCheckParameterClass")
+    @Override
+    public boolean equals(Object other) {
+        return sameClassAs(other) &&
+                equalsTo(getClass().cast(other));
+    }
+
+    private boolean equalsTo(TermStatQuery other) {
+        return Objects.equals(expr, other.expr)
+                && Objects.equals(aggr, other.aggr)
+                && Objects.equals(posAggr, other.posAggr)
+                && Objects.equals(terms, other.terms);
+    }
+
+    @Override
+    public Query rewrite(IndexReader reader) throws IOException {
+        return this;
+    }
+
+    @Override
+    public int hashCode() { return Objects.hash(expr, aggr, posAggr, terms); }
+
+    @Override
+    public String toString(String field) {
+        return null;
+    }
+
+    @Override
+    public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost)
+            throws IOException {
+        assert scoreMode.needsScores() : "Should not be used in filtering mode";
+
+        return new TermStatWeight(searcher, this, terms, scoreMode, aggr, posAggr);
+    }
+
+    static class TermStatWeight extends Weight {
+        private final String expression;
+        private IndexSearcher searcher;
+        private final Set<Term> terms;
+        private final ScoreMode scoreMode;
+
+        private AggrType aggr;
+        private AggrType posAggr;
+
+        TermStatWeight(IndexSearcher searcher, TermStatQuery query, Set<Term> terms, ScoreMode scoreMode, AggrType aggr, AggrType posAggr) {
+            super(query);
+            this.searcher = searcher;
+            this.expression = query.expr;
+            this.terms = terms;
+            this.scoreMode = scoreMode;
+            this.aggr = aggr;
+            this.posAggr = posAggr;
+        }
+
+        @Override
+        public void extractTerms(Set<Term> terms) { terms.addAll(terms); }
+
+        @Override
+        public Explanation explain(LeafReaderContext context, int doc) throws IOException {
+            Scorer scorer = this.scorer(context);
+            int newDoc = scorer.iterator().advance(doc);
+            if (newDoc == doc) {
+                return Explanation
+                        .match(scorer.score(), "weight(" + this.expression + " in doc " + newDoc + ")");
+            }
+            return Explanation.noMatch("no matching term");
+        }
+
+        @Override
+        public Scorer scorer(LeafReaderContext context) throws IOException {
+            Expression compiledExpression = (Expression) Scripting.compile(this.expression);
```

Comment:
you might perhaps want to compile this earlier (in TermStatQueryBuilder and pass Expression compiledExpression in the constructor of this query) so that it's built once per query as opposed to once per segment.

Reply:
Done.